src/essreflectometry/io.py

```diff
@@ -1,41 +1,117 @@
+    *,
+    filename: Union[str, os.PathLike, TextIO],
+    data: Iterable[sc.Variable],
+    orso: Orso,
 ) -> None:
     """
-    Save a data array with the ORSO .ort file format.
+    Save a data with the ORSO .ort file format.
 
     Parameters
     ----------
-    data_array:
-        Scipp-data array to save.
     filename:
-        Filename.
-    dimension:
-        String for dimension to perform mean over.
+        The file or filename to save to.
+    data:
+        Columns to save.
+        Must match the columns stored in ``orso`.columns``.
+    orso:
+        ORSO object holding the metadata for ``data``.
     """
-    from orsopy import fileio
-
-    if filename[:-4] == '.ort':
-        raise ValueError("The expected output file ending is .ort.")
-    if dimension is not None:
-        data_array = data_array.mean(dimension)
-    q = data_array.coords['Q']
-    if data_array.coords.is_edges('Q'):
-        q = sc.midpoints(q)
-    R = data_array.data
-    sR = sc.stddevs(data_array.data)
-    sq = data_array.coords['sigma_Q']
-    dataset = fileio.orso.OrsoDataset(
-        data_array.attrs['orso'].value,
-        np.array([q.values, R.values, sR.values, sq.values]).T,
+    _require_ort_suffix(filename)
+    orso = _validate_orso_header(orso)
+    dataset = OrsoDataset(
+        orso, np.column_stack([_extract_values_array(d) for d in data])
     )
-    fileio.orso.save_orso([dataset], filename)
+    dataset.save(filename)
+
+
+def save_iofq_ort(
+    filename: Union[str, os.PathLike, TextIO],
+    iofq: NormalizedIofQ,
+    data_source: OrsoDataSource,
+    reduction: OrsoReduction,
+    *,
+    sigma_q: Optional[QResolution] = None,
+) -> None:
+    """Save reduced I-of-Q data to an ORSO .ort file.
+
+    Parameters
+    ----------
+    filename:
+        The file or filename to save to.
+    iofq:
+        Reduced :math:`I(Q)`.
+    data_source:
+        ORSO ``DataSource`` with metadata about the data that
+        ``iofq`` was computed from.
+    reduction:
+        ORSO ``Reduction`` with metadata about the reduction process.
+    sigma_q:
+        Resolution of :math:`Q`.
+        If ``None``, ``iofq`` must have a coordinate called ``'sigma_Q'``
+        that stores the resolution.
+    """
+    orso = Orso(
+        data_source=data_source,
+        reduction=reduction,
+        columns=[
+            Column('Qz', '1/angstrom', 'wavevector transfer'),
+            Column('R', None, 'reflectivity'),
+            Column('sR', None, 'standard deviation of reflectivity'),
+            Column(
+                'sQz',
+                '1/angstrom',
+                'standard deviation of wavevector transfer resolution',
+            ),
+        ],
+    )
+
+    qz = iofq.coords['Q'].to(unit='1/angstrom', copy=False)
+    if iofq.coords.is_edges('Q'):
+        qz = sc.midpoints(qz)
+    r = sc.values(iofq.data)
+    sr = sc.stddevs(iofq.data)
+    if sigma_q is not None:
```

Comment:
Why do we need two cases here? Wouldn't it be easier to just either get  `sigma_Q` from the data or from an extra argument to this provider?

Reply:
It would be. And I'm going to write a provider that creates an OrsoDataset which will then request all it needs.